Support editor functions after reattach to DOM

sculpt0r opened the pull request:
<!--
🚨 If you want to submit a PR for a security vulnerability, please contact us directly
at https://ckeditor.com/contact/ instead. 🚨
-->
## What is the purpose of this pull request?

New feature

## Does your PR contain necessary tests?

All patches that change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [X] Unit tests
- [X] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [X] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
* [#4462](https://github.com/ckeditor/ckeditor4/issues/4462): Support editor functions after reattaching to DOM.
```

## What changes did you make?

- Used `MutationObserver` to detect whenever the editor is attached to DOM again.
- Based on that information, recreate the entire editable section. In the case of IE, I have to wait until `load` event. So only for IE, there is an extra flag used to mark, that editable should be recreated.
- Added extra flag which preserves `iframe` element during editable `detach` process. Used only internally in recreating.
- Added config option (`detachableParent`) which allows pass native DOM object which should be observed for detach / attach changes
- Added new `editor.status` (`recreating`) which prevents null references in recreating the process
- Added manual & unit tests

We won't be able to support IE below `11`, because of `MutationObserver` usage. However, it's mostly for integrations, which doesn't support IE below `11`.

Three units test are passing regardless of this fix:
- 'test reattached editor makes editor data empty with observed invalid DOM object'
- 'test reattached editor parent makes editor data empty with observed invalid DOM object'
- 'test reattached editor parent makes editor data empty with observed detached parent element'

They confirm if iframe reloading still makes the editor empty. Also, they show that an invalid configuration still leads to an empty editor.

Fix two tests that create query strings for `querySelector()` with restricted characters (IE):
- https://github.com/ckeditor/ckeditor4/pull/4601/files#diff-a15eeadc2752af4c0b9dcfc057fb5fe450dfb6cf9a672e4fc9baacaac9036903
- https://github.com/ckeditor/ckeditor4/pull/4601/files#diff-64582745fd35627ba47ea103cea4f90242ccede20a04a22c9c7311393b50dba7

## Which issues does your PR resolve?

Closes #4462

sculpt0r commented:
@ckeditor/qa-team Hello there :wave:   
Could you look at this feature again? The previous solution (#4481) was discarded and we came up with something better.

**Note** that it won't work for IE below `11`.

There is a manual test with some additional plugins (`tests/plugins/wysiwygarea/manual/detachedwithplugins`) which may be helpful during some monkey clicking.

LukaszGudel commented:
I've tested this feature using provided manual test and everything seems to work correctly 👍  I did not find anything broken during my testing.

Dumluregn commented on plugins/wysiwygarea/plugin.js:
```diff
@@ -716,3 +796,14 @@ CKEDITOR.config.disableNativeSpellChecker = true;
+ * Editor `wysiwygarea` iframe will be recreated whenever this element will be reattached to DOM.
+ *
+ * @since 4.17.0
+ * @cfg {Object} [detachableParent=CKEDITOR.document.$]
```
Well, it is an object in the JS understanding, but it doesn't distinguish itself from non-DOM objects, so I'd put [`HTMLElement`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement) here - it's interface and not exactly a type, but it's much more informative (it's the same case as with `Array` or `RegExp`).

Dumluregn commented on tests/plugins/wysiwygarea/detached.js:
```diff
@@ -0,0 +1,260 @@
+				assert.ignore();
+			}
+
+			var startupData = '<p>CKEditor4</p>';
```
AFAICS this data is the same for each test, so we could define it before and reuse it.

Dumluregn commented on tests/plugins/wysiwygarea/detached.js:
```diff
@@ -0,0 +1,260 @@
+				CKEDITOR.tools.setTimeout( function() {
+					resume( function() {
+						var iframeElement = bot.editor.ui.space( 'contents' ).findOne( 'iframe' ),
+							editorData = iframeElement.getFrameDocument().getBody().getHtml();
```
Can't we just use `bot.editor.getData()`?

sculpt0r replied:
Sadly - we cannot :cry: It involves internal logic and it returns 'valid' data even if the iframe itself reloads and has no content. So with `bot.editor.getData()` tests could pass or fail due to wrong reasons.

For example `test reattached editor contains the same data with observed default dom object` can pass even without the fix. OTOH test that expects empty editor fails, because of `getData()`.

Dumluregn commented on tests/plugins/wysiwygarea/detached.js:
```diff
@@ -0,0 +1,260 @@
+				startupData: startupData,
+				name: 'editor' + new Date().getTime(),
+				config: {
+					extraAllowedContent: 'iframe'
```
I didn't notice where we use it? Shouldn't we provide custom startupData containing iframe in this test?

Dumluregn replied:
Actually I bet you wanted to use fixture with `editor5` here as it contains `iframe` and is not used anywhere?

Dumluregn reviewed (changes requested):
As for the solution, it looks much clearer now with `mutationObserver` (no custom attributes on iframe native element, yay!). It looks to work just like before 👍🏻 

Besides inline comments, I'd like to ask for explicit tests for the cases that were [failing in the previous version of the solution](https://github.com/ckeditor/ckeditor4/pull/4481#pullrequestreview-616530029), so the embed test and an image with undo test (and BTW I see the second one is still failing - could you relate to it since you didn't in the original PR?).

sculpt0r commented on plugins/wysiwygarea/plugin.js:
```diff
@@ -93,9 +98,78 @@
+
+					// (#4462)
+					var isIE11 = CKEDITOR.env.ie && !CKEDITOR.env.edge && CKEDITOR.env.version === 11;
+					if ( isIE11 ) {
```
In IE11 we have to wait until `load` event, before `resume()` will be called. If we don't - there is an error:

https://github.com/ckeditor/ckeditor4/blob/7481c473b85b14f0bbfcd24fb5b368ebbc1fc3f9/core/dom/element.js#L1264-L1267

Which says: `SCRIPT5007: Unable to get property 'setAttribute' of undefined or null reference`. In this case, we try to set `contenteditable=true` on an invalid reference. So `this.$` is invalid and normally it should point to the body of the iframe. But iframe just lost its content, so there is no `body` inside. In other browsers there is a body - it's empty - but the primary document is available.

sculpt0r commented on plugins/wysiwygarea/plugin.js:
```diff
@@ -93,9 +98,78 @@
+					}
+
+					// All non IE browsers or IE11.
+					if ( !( CKEDITOR.env.ie && !CKEDITOR.env.edge ) || ( isIE11 ) ) {
```
I thought about it, but double IE11 check looks not so good to me. But overall - indeed looks like one part of some fix :+1:

sculpt0r commented on tests/plugins/wysiwygarea/detached.js:
```diff
@@ -0,0 +1,260 @@
+
+	bender.test( {
+		'test reattached editor contains the same data with observed default dom object': function() {
+			if ( !isSupportedEnvironment() ) {
```
:thinking: Verify environment explicit in each test looks quite redundant, OTOH - you don't have to scroll all of the tests to find if one of them is ignored. And - its hard to compare those long tests titles (at least for me ;p). So I will stay with the current solution.

sculpt0r commented on tests/plugins/wysiwygarea/manual/detachedwithplugins.html:
```diff
@@ -0,0 +1,43 @@
+		editorDetached = false;
+
+	setButton.on( 'click', function() {
+		// debugger
```
:see_no_evil:

sculpt0r commented on plugins/wysiwygarea/plugin.js:
```diff
@@ -716,3 +796,14 @@ CKEDITOR.config.disableNativeSpellChecker = true;
+/**
+ * Editor detachable parent. Native DOM object.
+ *
+ * Editor `wysiwygarea` iframe will be recreated whenever this element will be reattached to DOM.
```
I change the name to `observableParent`. Because we can't observe element which is detachable. If so, then no changes will be detected. We have to choose at least the parent of the detachable element.

sculpt0r commented:
@Dumluregn - ready for review.

- added missing tests from the previous PR
- correct inline comments

I try to fix this additional step in history - but without success... TBH I was able to fix it, but at the same time `undo` was broken if I toggle editor before insert image - so still wrong behaviour - even worse. As you write - could we extract it as another issue? But in this case, we will have one failing manual test.

Dumluregn commented on plugins/wysiwygarea/plugin.js:
```diff
@@ -93,9 +98,82 @@
+					}
+				}
+
+				function attachIframeReloader( evt ) {
+					evt && evt.removeListener();
+
+					iframe.on( 'load', function() {
+						if ( recreateEditable ) {
+							recreateEditable = false;
+							recreate();
+						}
+					} );
+				}
```
This function is actually invoked before `observeEditor()` etc, so let's move the declaration higher to be easier to find.

Dumluregn commented on plugins/wysiwygarea/plugin.js:
```diff
@@ -93,9 +98,82 @@
+					if ( !node.querySelector ) {
+						return;
+					}
+					var editorSelector = '#' + editor.container.getId(),
```
```suggestion

					var editorSelector = '#' + editor.container.getId(),
```

Dumluregn commented on plugins/wysiwygarea/plugin.js:
```diff
@@ -93,9 +98,82 @@
+
+					// (#4462)
+					// Skip IE's below version 11. They don't support MutationObserver.
+					if ( !CKEDITOR.env.ie || isIE11 ) {
+						if ( isIE11 ) {
+							editor.on( 'mode', attachIframeReloader, { iframe: iframe, editor: editor, callback: callback } );
+						}
+
+						observeEditor();
+					}
```
We could flatten it even more:
```suggestion
					if ( CKEDITOR.env.ie && !isIE11 ) {
						return;
					}
					
					if ( isIE11 ) {
						editor.on( 'mode', attachIframeReloader, { iframe: iframe, editor: editor, callback: callback } );
					}

					observeEditor();
```

But the conditions now are quite clear so we can also leave it as is if you want 🤔

sculpt0r replied:
Well, it looks even better - so yes - lets make it flat :D

Dumluregn commented on plugins/wysiwygarea/plugin.js:
```diff
@@ -716,3 +800,33 @@ CKEDITOR.config.disableNativeSpellChecker = true;
+ * * make sure detachable element is placed any nested depth under observable element.
+ * * make sure editor is placed any nested depth under detachable element.
+ *
+ * See overview of HTML hierarchy below:
```
I think it's important to mention that it's just a pseudocode and not actual HTML tags 🙃

Dumluregn reviewed (changes requested):
Just some small refactors left.

Besides that - on IE11 I've encountered an error after each toggle:
![Screenshot 2021-04-15 at 15 48 38](https://user-images.githubusercontent.com/10883184/115002026-8f3e9300-9ea4-11eb-9976-559393e0f07f.png)

Looks like Bender issue TBH 🤔  could you verify if it's reproducible outside of testing environment? I didn't experience any influence on editor (all tests pass), but in manual descriptions we write that any error causes the test to fail, so we should take some action.

sculpt0r commented on plugins/wysiwygarea/plugin.js:
```diff
@@ -716,3 +800,33 @@ CKEDITOR.config.disableNativeSpellChecker = true;
+ *
+ * To recreate editor `wysiwygarea` iframe after reattach editor to DOM:
+ *
+ * * make sure detachable element is placed any nested depth under observable element.
+ * * make sure editor is placed any nested depth under detachable element.
```
I meant that those elements don't have to be direct children. So your observable could be a top-level `div` element and the editor itself could be placed deep down under many others elements.

Dumluregn replied:
Ok, then I guess `is nested on any level` should be a little easier to understand 🤔

sculpt0r commented:
Rebase on newest major

sculpt0r commented:
> Besides that - on IE11 I've encountered an error after each toggle:
> ![Screenshot 2021-04-15 at 15 48 38](https://user-images.githubusercontent.com/10883184/115002026-8f3e9300-9ea4-11eb-9976-559393e0f07f.png)

Definitely, it's related to our solution. Because in unit tests which expect empty editor (so didn't call restore logic) this issue doesn't appear for me in IE11. All other units from `detached` throw this error... :thinking:

sculpt0r commented:
@Dumluregn - ready for review.
- rebased on the newest major
- correct all inline comments
- Suppress IE11 error - please take a look at https://github.com/ckeditor/ckeditor4/pull/4601#issuecomment-824855508

Dumluregn reviewed (changes requested):
I've encountered an issue with undo history which occurs on all browsers and is rather simple to reproduce, so we have to take care of it.

1. Open any manual test.
2. Toggle source mode.
3. Toggle editor (detach+reattach).
4. Use `bold` button.

Editor doesn't record any undo step. Even if it looks like it does (after you type anything), you can't access it. 

Other than that - mostly small ammendments.

sculpt0r commented:
Rebase on newest major.

sculpt0r commented on tests/plugins/wysiwygarea/manual/detachedembededimage.md:
```diff
@@ -0,0 +1,13 @@
+1. Use 'Insert Media Embed' and put there any image address. For example google logo.
+2. Reattach editor with double click on "Toggle" button.
+
+	**Expected:** There is exactly one undo step.
```
Yes. Still we have a bug here: https://github.com/ckeditor/ckeditor4/pull/4601#issuecomment-818285341

sculpt0r commented:
> I've encountered an issue with undo history which occurs on all browsers and is rather simple to reproduce, so we have to take care of it.
> 
> 1. Open any manual test.
> 2. Toggle source mode.
> 3. Toggle editor (detach+reattach).
> 4. Use `bold` button.
> 
> Editor doesn't record any undo step. Even if it looks like it does (after you type anything), you can't access it.

I've got this one! :D

sculpt0r commented:
Probably have some explanation for doubled history, but I'll put in in extracted issue.

sculpt0r commented on tests/plugins/wysiwygarea/detached.js:
```diff
@@ -0,0 +1,206 @@
+
+			bender.editorBot.create( {
+				name: 'editor',
+				startupData: startupData,
```
We use `startupData` in assertions. That means, if we put it in `.html` file, we need to add additional data extraction from HTML. I would stay with the current solution. Actually, current HTML is provided for hierarchy visualization.

Dumluregn replied:
Yeah, we have to keep `startupData` variable for assertions, but I don't see why we'd need to extract data from HTML?

Dumluregn replied:
I agreed with @sculpt0r that having the data in variable is less prone to produce errors in case of changes in the expected data or test structure, so we'll leave it as is 👍🏻

sculpt0r commented:
@Dumluregn - ready, steady. review! :D
- Sorry for 📄the wall of text below...
- Correct all inline comments, besides https://github.com/ckeditor/ckeditor4/pull/4601#discussion_r619112750. Please take a look at my answer: https://github.com/ckeditor/ckeditor4/pull/4601#discussion_r620910525
- In the case of [broken undo history](https://github.com/ckeditor/ckeditor4/pull/4601#discussion_r619119094), I extracted a separate issue: https://github.com/ckeditor/ckeditor4/issues/4644, correct manual test scenario - so it passes at the current state and I add a note about this issue in the test.
- I moved to modern `editor.setData` syntax: https://github.com/ckeditor/ckeditor4/blob/0c132bfc377f7d051c913f327519f8ff815d8bce/plugins/wysiwygarea/plugin.js#L182. However, I switch default `fireSnapshot` to `false` - so no additional snapshots would be taken during recreation. undo plugin still fails, but we don't need those snapshots at recreating time.

- The most important, but also most complex: https://github.com/ckeditor/ckeditor4/pull/4601#pullrequestreview-643147098. The reason for that was invalid `editor.status`.
  - Just before `setData` https://github.com/ckeditor/ckeditor4/blob/0c132bfc377f7d051c913f327519f8ff815d8bce/plugins/wysiwygarea/plugin.js#L181-L182 we are changing it to `recreating`. It is because undo plugin early returns from saving snapshots with status different than `ready` (BTW, I try with `lockSnapshot` event, which causes below `this.locked` to be true, but this makes the entire data gone...) https://github.com/ckeditor/ckeditor4/blob/0c132bfc377f7d051c913f327519f8ff815d8bce/plugins/undo/plugin.js#L372 We are not setting status to `ready` after `SetData` in wysiwygmode plugin. Because when the editor is created, the callback is passed and stays as context for our `recreate` function. So each `recreate` uses a callback which sets editor status back to `ready`. https://github.com/ckeditor/ckeditor4/blob/0c132bfc377f7d051c913f327519f8ff815d8bce/core/creators/themedui.js#L378-L386 However, if we are getting back from source mode, there is another `setMode` invoked without any callback. Now our `recreate` doesn't has a previous callback context, so nothing changes the status back. https://github.com/ckeditor/ckeditor4/blob/0c132bfc377f7d051c913f327519f8ff815d8bce/plugins/sourcearea/plugin.js#L145 Also we are not taking status back in `recreate` right after the `editor.setData` becuase it will be to early. In such case we are getting back to not working editor. `editor.status = 'ready'` which comes from `themedui.js` is set in timeout with zero-time, so after all stack invocations. Based on that, I decide to bind again to `mode` event https://github.com/ckeditor/ckeditor4/blob/0c132bfc377f7d051c913f327519f8ff815d8bce/plugins/wysiwygarea/plugin.js#L117-L119 and change status to `ready` again. This `mode` event is fired in a similar way that editor creation, but it is placed in a method which is called to change mode: https://github.com/ckeditor/ckeditor4/blob/0c132bfc377f7d051c913f327519f8ff815d8bce/core/creators/themedui.js#L245.

Dumluregn commented on plugins/wysiwygarea/plugin.js:
````diff
@@ -716,3 +808,33 @@ CKEDITOR.config.disableNativeSpellChecker = true;
+ * See overview of hierarchy below (HTML pseudocode):
+ *
+ * ```
+ * <observable>
+ *   <detachable> <!-- Any nesting depth -->
+ *     <editor></editor> <!-- Any nesting depth -->
+ *   </detachable>
+ * </observable>
````
Maybe adding `...` instead of the comment?

```suggestion
 * <observable>
 *    <...>
 *     <detachable>
 *       <...>
 *         <editor></editor>
 *       </...>
 *     </detachable>
 *    </...>
 * </observable>
```

Dumluregn commented on plugins/wysiwygarea/plugin.js:
````diff
@@ -716,3 +808,33 @@ CKEDITOR.config.disableNativeSpellChecker = true;
+ * </observable>
+ * ```
+ *
+ * By default, the entire document is observed.
+ *
+ * If you know exactly which element is detaching - choose its direct parent.
````
```suggestion
 * By default, the entire document is observed. However, if you know exactly which element is detaching, you can choose its direct parent to increase performance a little.
```

Dumluregn commented on plugins/wysiwygarea/plugin.js:
```diff
@@ -716,3 +808,33 @@ CKEDITOR.config.disableNativeSpellChecker = true;
+ *
+ * If you know exactly which element is detaching - choose its direct parent.
+ *
+ * Note, that if you choose element which is detaching. No changes will be detected.
```
```suggestion
 * Note, that if you choose element which is detaching, no changes will be detected.
```

Dumluregn commented on plugins/wysiwygarea/plugin.js:
```diff
@@ -716,3 +808,33 @@ CKEDITOR.config.disableNativeSpellChecker = true;
  */
+
+/**
+ * Document observation starting point. Native DOM object.
```
The first line should say something about the purpose of this option, so I'd add a few words like `mutation observer` and `detach` 😛

Dumluregn reviewed (changes requested):
> Based on that, I decide to bind again to mode event

Works fine now 👍🏻 

The PR looks good... enough to start renamings and polishing docs 📝 🔥 But it shouldn't take much effort.

Also please add a manual test with the scenario that was failing before (https://github.com/ckeditor/ckeditor4/pull/4601#pullrequestreview-643147098). And I've commented here https://github.com/ckeditor/ckeditor4/pull/4601#discussion_r621319217 (linking since it's easy to omit).

sculpt0r commented:
@Dumluregn - ready for another review.
- use proper name for `setData` call. I just change the name. It still has the same default value (which evaluate as `false`). So technically we could skip it, but I want to preserve it in this case, since there is a lot going on with `undoplugin` and we still have one extracted issue about it.
- Correct API doc comments and other smaller things according to your suggestions
- Added a new manual test (as you requested) with a previously failing case. I remove the note with the console from there because there is no need for additional checks here. Results are visible right in editor UI.

Dumluregn commented:
Rebased onto latest `major`, PR is almost ready for me. One last thing I've noticed is duplicated undo step (again), but this time on FF only:

1) Apply bold to text in editor
2) Detach and reattach editor
3) Check undo steps; should be 1, but there are 2.

Since this doesn't break anything and is browser specific, please take a look at it but don't spend more than 1 hour trying to find the cause and if you don't manage, we'll extract an issue. 

> Besides that, my sonarqube plugin shows that this invocation doesn't need an argument. But since there is a lot going on, I don't touch it yet. Do you think we could remove this argument now?

Normally I'd say it's better not to touch it, but here it looks like a very simple situation where the function was nested to use the scoped `editor` variable but it wasn't removed from arguments list, so I did remove it 👍🏻 

No more comments from me 👍🏻

Dumluregn reviewed (changes requested):
Rebased onto latest `major`, PR is almost ready for me. One last thing I've noticed is duplicated undo step (again), but this time on FF only:

1) Apply bold to text in editor
2) Detach and reattach editor
3) Check undo steps; should be 1, but there are 2.

Since this doesn't break anything and is browser specific, please take a look at it but don't spend more than 1 hour trying to find the cause and if you don't manage, we'll extract an issue. 

> Besides that, my sonarqube plugin shows that this invocation doesn't need an argument. But since there is a lot going on, I don't touch it yet. Do you think we could remove this argument now?

Normally I'd say it's better not to touch it, but here it looks like a very simple situation where the function was nested to use the scoped `editor` variable but it wasn't removed from arguments list, so I did remove it 👍🏻 

No more comments from me 👍🏻

sculpt0r commented:
@Dumluregn - done ;)

I track a little for this bug, but finally decided to extract it (https://github.com/ckeditor/ckeditor4/issues/4667) and describe what I found. And as we both know, there is no easy and obvious situation here. What is more, this happens not only for FF for me.

Thank you for your corrects and for the fix with this unused function argument.

Dumluregn reviewed (approved):
LGTM! 🎉 

Your turn, @f1ames ... 😄

f1ames commented:
> Your turn, @f1ames ...

I think I will ask @Comandeer to take this review :smile:

Comandeer commented:
Rebased onto latest `major`.

Comandeer commented on tests/plugins/wysiwygarea/manual/detachedembededimagesource.html:
```diff
@@ -0,0 +1,26 @@
+<p>Google logo:</p>
+
+<pre>https://www.google.com/images/branding/googlelogo/1x/googlelogo_color_272x92dp.png</pre>
```
I'm not sure if we should use some resource that we don't have any control over.

Comandeer commented on tests/plugins/wysiwygarea/manual/detachedtwoeditors.html:
```diff
@@ -0,0 +1,38 @@
+<button id="toggleButton">Toggle</button>
```
I'd move this button directly above detachable editor. It seems more intuitive for me. Additionally, currently I need to scroll page up to use the button after interacting with the second editor.

Comandeer commented on tests/tickets/4527/1.js:
```diff
@@ -7,7 +7,6 @@
 	bender.editor = {
 		editorName: 'editor1',
 		startupData: '',
-		name: document.title,
```
Why is this change introduced?

sculpt0r replied:
This one leads to errors in IE. This name was used inside `querySelector`. Unfortunately in IE, there were some restricted characters in the selector. The same goes for https://github.com/ckeditor/ckeditor4/pull/4601/files#diff-a15eeadc2752af4c0b9dcfc057fb5fe450dfb6cf9a672e4fc9baacaac9036903

So both tests start to fail because of this.

Comandeer reviewed (changes requested):
I've discovered some weird bug connected with undo:

1. Open http://localhost:1030/tests/plugins/wysiwygarea/manual/detachedsourceundo
2. Bold text.
3. Undo.
4. Press "Toggle" button twice.
5. Focus the editable.
6. Switch to source mode.
7. Press "Toggle" button twice.
8. Switch back to WYSIWYG mode.

The redo step is lost.

There aren't also any unit tests for undo after reattaching.

sculpt0r commented:
Rebase on newest major.

sculpt0r commented:
@Comandeer - ready for review
- correct all your inline comments. 
- add an explanation for your question [here](https://github.com/ckeditor/ckeditor4/pull/4601#discussion_r628106744)

sculpt0r commented on plugins/wysiwygarea/plugin.js:
```diff
@@ -93,9 +103,79 @@
+						return;
+					}
+
+					if ( isIE11 ) {
+						recreateEditable = true;
+					} else {
+						recreate();
+					}
```
Well, that indicates that `recreate()` returns some meaningful value - and it's not. We know that will be undefined but if you read it looks like recreation could end with success or failure.
So to make consensus, maybe:

```js
if ( !isIE11 ) {
  recreate();
  return;
}

recreateEditable = true;
```

?

Comandeer reviewed (changes requested):
Looks good! But I see that some error appeared in IE11 due to new tests:
<img width="518" alt="Permission denied error before redo test" src="https://user-images.githubusercontent.com/1078728/118465575-4deb0e80-b702-11eb-84ca-426209b11c24.png">
Could you check it?

sculpt0r commented:
@Comandeer - ready for review
I found that access to document while editable is recreating caused this issue. So I simply prevent that and return null: https://github.com/ckeditor/ckeditor4/pull/4601/commits/b9544f37196ef34904b2468b5c23ad7b19f634f8

Comandeer commented on tests/plugins/wysiwygarea/detached.js:
```diff
@@ -0,0 +1,263 @@
+			} );
+		},
+
+		'test reattached editor makes editor data empty with observed invalid DOM object': function() {
```
This and following tests fail on IE8.

sculpt0r replied:
It's funny - they failed for me only with opened dev console :thinking: And I can run them with dev console -> fail. Then close this console and they pass. So it doesn't look like cache :thinking:

Comandeer replied:
Strange. For me they fail even without console.

Comandeer commented on core/selection.js:
```diff
@@ -1260,7 +1260,8 @@
 		var editable = this.editable();
-		return editable && this.mode == 'wysiwyg' ? new CKEDITOR.dom.selection( editable ) : null;
+
+		return editable && this.mode == 'wysiwyg' && editable.editor.status !== 'recreating' ? new CKEDITOR.dom.selection( editable ) : null;
```
It would be nice to add some test to really see if `null` is returned when editor is recreating.

Comandeer reviewed (changes requested):
Almost there… See comments.

sculpt0r commented:
@Comandeer - ready for another review.

- Added unit test for getSelection()
- Fix failing tests. They failed because the previous test has still an active observer (the default observed element is the entire document).  So now - the editor is destroyed at the end of the test.

Comandeer reviewed (approved):
LGTM! 🎉 

I've just updated API docs for `CKEDITOR.editor#getSelection()` (it seems that we have them wrong for years 🙈) and simplify it a little bit (replaced `editable.editor.status` with `this.status`).

I've also added additional changelog entry for the change in `CKEDITOR.editor#getSelection()` under API Changes.
